src/dmd/e2ir.d

```diff
@@ -2758,30 +2774,74 @@ elem *toElem(Expression e, IRState *irs)
 
                     assert(ae.e2.type.ty != Tpointer);
 
-                    if (!postblit && !destructor && !irs.arrayBoundsCheck())
+                    if (!postblit && !destructor)
                     {
                         elem *ex = el_same(&eto);
 
-                        // Determine if elen is a constant
-                        elem *elen;
-                        if (eto.Eoper == OPpair &&
-                            eto.EV.E1.Eoper == OPconst)
+                        /* Returns: length of array ex
+                         */
+                        static elem *getDotLength(elem *eto, elem *ex)
+                        {
+                            if (eto.Eoper == OPpair &&
+                                eto.EV.E1.Eoper == OPconst)
+                            {
+                                // It's a constant, so just pull it from eto
+                                return el_copytree(eto.EV.E1);
+                            }
+                            else
+                            {
+                                // It's not a constant, so pull it from the dynamic array
+                                return el_una(global.params.is64bit ? OP128_64 : OP64_32, TYsize_t, el_copytree(ex));
+                            }
+                        }
+
+                        auto elen = getDotLength(eto, ex);
+                        auto nbytes = el_bin(OPmul, TYsize_t, elen, esize);  // number of bytes to memcpy
+                        auto epto = array_toPtr(ae.e1.type, ex);
+
+                        elem *epfr;
+                        elem *echeck;
+                        if (irs.arrayBoundsCheck()) // check array lengths match and do not overlap
                         {
-                            elen = el_copytree(eto.EV.E1);
+                            auto ey = el_same(&efrom);
+                            auto eleny = getDotLength(efrom, ey);
+                            epfr = array_toPtr(ae.e2.type, ey);
+
+                            // length check: (eleny == elen)
+                            auto c = el_bin(OPeqeq, TYint, eleny, el_copytree(elen));
+
+                            /* Don't check overlap if epto and epfr point to different symbols
+                             */
```

Comment:
This appears to be vulnerable to pointer aliasing (epto and epfr could be pointing to arguments of an enclosing function, right?).

Reply:
Backend symbols themselves are always distinct, so no.